floodlight/core/events.py

```diff
@@ -61,3 +95,148 @@ def custom(self):
+        ]
+
+        if not missing_columns:
+            return None
+        else:
+            return missing_columns
+
+    @property
+    def essential_invalid(self):
+        incorrect_columns = [
+            col
+            for col in self.essential
+            if not self.column_values_in_range(col, essential_events_columns)
+        ]
+
+        if not incorrect_columns:
+            return None
+        else:
+            return incorrect_columns
+
+    @property
+    def protected_missing(self):
+        missing_columns = [
+            col for col in protected_columns if col not in self.protected
+        ]
+
+        if not missing_columns:
+            missing_columns = None
+
+        return missing_columns
+
+    @property
+    def protected_invalid(self):
+        incorrect_columns = [
+            col
+            for col in self.protected
+            if not self.column_values_in_range(col, protected_columns)
+        ]
+        if not incorrect_columns:
+            incorrect_columns = None
+
+        return incorrect_columns
+
+    def column_values_in_range(self, col: str, definitions: Dict[str, Dict]) -> bool:
+        """Perform the checks for value range for a single column of the inner event
+        DataFrame using the specifications from floodlight.core.definitions.
+
+        Parameters
+        ----------
+        col: str
+            Column of the inner event DataFrame to be checked
+        definitions: Dict
+            Dictionary for each column name to be checked. Each entry is a Dictionary
+            containing definitions for allowed value ranges of the respective
+            column. Information about value_range is given as a List of the form
+            [min_value, max_value].
+
+        Returns
+        -------
+        bool
+            True if the checks for value range pass and False otherwise
+        """
+        # skip if value range is not defined
+        if definitions[col]["value_range"] is None:
+            return True
+
+        # skip values that are None or NaN
+        check_col = self.events[col].dropna()
+
+        # check value range for remaining values
+        if not (definitions[col]["value_range"][0] <= check_col).all():
+            return False
+        if not (check_col <= definitions[col]["value_range"][1]).all():
+            return False
+
+        # all checks passed
+        return True
+
+    def add_frameclock(self, framerate: int):
+        """Add the column "frameclock", computed as the rounded multiplication of
+        gameclock and framerate, to the inner events DataFrame.
+
+        Parameters
+        ----------
+        framerate: int
+            Temporal resolution of data in frames per second/Hertz.
+        """
+        frameclock = np.full((len(self.events)), -1, dtype=int)
+        frameclock[:] = np.floor(self.events["gameclock"].values * framerate)
+        self.events["frameclock"] = frameclock
+
+    def find(
```

Comment:
After reviewing, I really feel that filter is a better name. As this might be confused with the inbuilt filter method, maybe something like ``filter_by``? Or ``select``?

Reply:
Chose select here.